File: src/main/java/com/leetcode/solutions/stack/validparentheses/ValidParentheses.java
package com.leetcode.solutions.stack.validparentheses;

import java.util.Stack;

public class ValidParentheses {
    public static boolean isValid(String s) {
        // Stack to keep track of opening brackets
        Stack<Character> stack = new Stack<>();

        // Mapping of closing to opening brackets
        for (char ch : s.toCharArray()) {
            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch); // Push opening brackets onto stack
            } else {
                if (stack.isEmpty()) return false; // No matching opening bracket

                char top = stack.pop(); // Get last opened bracket

                // Check if it matches the correct opening bracket
                if ((ch == ')' && top != '(') ||
                        (ch == '}' && top != '{') ||
                        (ch == ']' && top != '[')) {
                    return false;
                }
            }
        }

        return stack.isEmpty(); // Stack should be empty if valid
    }

    public static void main(String[] args) {
        System.out.println(isValid("()")); // true
        System.out.println(isValid("()[]{}")); // true
        System.out.println(isValid("(]")); // false
        System.out.println(isValid("([])")); // true
        System.out.println(isValid("([)]")); // false
    }
}

File: src/main/kotlin/com/leetcode/solutions/stack/validparentheses/ValidParentheses.kt
package com.leetcode.solutions.stack.validparentheses

fun isValid(s: String): Boolean {
    val stack = ArrayDeque<Char>()
    val bracketMap = mapOf(')' to '(', '}' to '{', ']' to '[')

    for (char in s) {
        if (char in bracketMap.values) {
            // If it's an opening bracket, push to stack
            stack.addLast(char)
        } else if (char in bracketMap.keys) {
            // If it's a closing bracket, check stack
            if (stack.isEmpty() || stack.removeLast() != bracketMap[char]) {
                return false
            }
        }
    }

    return stack.isEmpty() // If stack is empty, all brackets matched correctly
}

fun main() {
    val testCases = listOf(
        "()", "()[]{}", "(]", "([])", "{[()]}", "{[(])}", ""
    )

    for (test in testCases) {
        println("Input: \"$test\" -> Output: ${isValid(test)}")
    }
}